contracts/EscrowAndStaking/CrossChainCollector.sol

```diff
@@ -0,0 +1,408 @@
+        gasLimitUsed = 1000000;
+        if (_masterCollector) {
+            masterCollector = address(this);
+            masterCollectorChain = _masterCollectorSelector;
+            chainSelector[0] = _masterCollectorSelector;
+            collectorAddress[_masterCollectorSelector] = address(this);
+            chainSelectorIndex[_masterCollectorSelector] = 0;
+            numOfActiveCollectors = 1;
+            period = 1;
+        }
+    }
+
+    receive() external payable {}
+
+    /* ========== VIEW FUNCTIONS ========== */
+
+    function isMasterCollector() external view returns (bool isMaster) {
+        isMaster = masterCollector == address(this);
+    }
+
+    function lastMessageFromChainSelector() external view returns (uint64 chainSelector_) {
+        if (numOfMessagesReceived > 0) {
+            chainSelector_ = messagesReceivedFromChainSelector[numOfMessagesReceived - 1];
+        }
+    }
+
+    /* ========== EXTERNAL FUNCTIONS ========== */
+
+    function sendOnClosePeriod(
+        uint _totalStakedLastPeriodEnd,
+        uint _totalEscrowedLastPeriodEnd,
+        uint _bonusPoints,
+        uint _revShare
+    ) external {
+        require(msg.sender == stakingThales, "InvSender");
+        if (masterCollector == address(this)) {
+            _storeRewards(
+                masterCollectorChain,
+                _totalStakedLastPeriodEnd,
+                _totalEscrowedLastPeriodEnd,
+                _bonusPoints,
+                _revShare
+            );
+            ++numOfCollectedResultsForThisPeriod;
+            if (numOfCollectedResultsForThisPeriod == numOfActiveCollectors) {
+                readyToBroadcast = true;
+            }
+        } else {
+            bytes memory message = abi.encode(
+                _totalStakedLastPeriodEnd,
+                _totalEscrowedLastPeriodEnd,
+                _bonusPoints,
+                _revShare
+            );
+            _sendMessageToChain(masterCollectorChain, message);
+        }
+        emit SentOnClosePeriod(_totalStakedLastPeriodEnd, _totalEscrowedLastPeriodEnd, _bonusPoints, _revShare);
+    }
+
+    function broadcastMessageToAll() external onlyOwner {
+        if (readyToBroadcast) {
+            _broadcastMessageToAll();
+            numOfCollectedResultsForThisPeriod = 0;
+            readyToBroadcast = false;
+            ++period;
+            if (weeklyRewardsDecreaseFactor > 0) {
+                _setRewardsForNextPeriod();
+            }
+        }
+    }
+
+    /* ========== INTERNAL FUNCTIONS ========== */
+
+    /// handle a received message
+    function _ccipReceive(Client.Any2EVMMessage memory any2EvmMessage) internal override {
+        // decoding the message
+        if (!messageIdAlreadyReceived[any2EvmMessage.messageId]) {
+            address sender = abi.decode(any2EvmMessage.sender, (address));
+            messagesReceivedFromChainSelector[numOfMessagesReceived] = any2EvmMessage.sourceChainSelector;
```

Comment:
why do we store received messages on chain in these mappings? Just wondering if its only for record keeping or actually has a use case?

Reply:
is `sourceChainSelector` smth that comes from CCIP?

Reply:
> why do we store received messages on chain in these mappings? Just wondering if its only for record keeping or actually has a use case?

the mappings are just record keeping, which can be removed

Reply:
they will remain kept due to the storage slots